types/wlr_xdg_foreign_v1.c

```diff
@@ -0,0 +1,458 @@
+
+	wl_list_insert(&foreign->exporter.objects, &exporter->link);
+}
+
+static struct wlr_xdg_importer_v1 *xdg_importer_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_importer_v1_interface,
+		&xdg_importer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_importer_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void xdg_imported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_imported_v1 *imported = xdg_imported_from_resource(resource);
+
+	struct wlr_xdg_imported_child_v1 *child, *child_tmp;
+	wl_list_for_each_safe(child, child_tmp, &imported->children, link) {
+		struct wlr_xdg_surface *xdg_child =
+			wlr_xdg_surface_from_wlr_surface(child->surface);
+		wlr_xdg_toplevel_set_parent(xdg_child, NULL);
+	}
+
+	wl_list_remove(&imported->exported_destroyed.link);
+	wl_list_remove(&imported->link);
+	free(imported);
+}
+
+static void xdg_imported_handle_exported_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_xdg_imported_v1 *imported =
+		wl_container_of(listener, imported, exported_destroyed);
+	disconnect_imported(imported);
+}
+
+static void xdg_importer_handle_import(struct wl_client *wl_client,
+				struct wl_resource *client_resource,
+				uint32_t id,
+				const char *handle) {
+	struct wlr_xdg_importer_v1 *importer =
+		xdg_importer_from_resource(client_resource);
+
+	struct wlr_xdg_imported_v1 *imported =
+		calloc(1, sizeof(struct wlr_xdg_imported_v1));
+	if (imported == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	imported->exported = wlr_xdg_foreign_registry_find_by_handle(
+		importer->foreign->registry, handle);
+	imported->resource = wl_resource_create(wl_client,
+		&zxdg_imported_v1_interface,
+		wl_resource_get_version(importer->resource),
+		id);
+	wl_resource_set_implementation(imported->resource, &xdg_imported_impl,
+			imported, xdg_imported_handle_resource_destroy);
+
+	wl_list_init(&imported->children);
+	wl_list_insert(&importer->imports, &imported->link);
+
+	if (imported->exported == NULL) {
+		wl_list_init(&imported->exported_destroyed.link);
+		zxdg_imported_v1_send_destroyed(imported->resource);
+	} else {
+		imported->exported_destroyed.notify = xdg_imported_handle_exported_destroy;
+		wl_signal_add(&imported->exported->events.destroy, &imported->exported_destroyed);
+	}
+}
+
+static const struct zxdg_importer_v1_interface xdg_importer_impl = {
+	.destroy = xdg_importer_handle_destroy,
+	.import = xdg_importer_handle_import
+};
+
+static void xdg_importer_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_importer_v1 *importer =
+		xdg_importer_from_resource(resource);
+	struct wlr_xdg_imported_v1 *imported, *imported_tmp;
+	wl_list_for_each_safe(imported, imported_tmp, &importer->imports, link) {
+		disconnect_imported(imported);
```

Comment:
Nothing in the protocol says that destroying the importer also destroys the child objects.